docs/src/appendix/packaging.md

```diff
@@ -0,0 +1,501 @@
+a package with a lower-versioned VM will fail. This field is optional, and when not present, is equivalent
+to no minimum required version.
+
+#### Package Type
+
+There are two primary types of packages: libraries and programs (executables). These correspond to the
+`[lib]` and `[bin]` sections of the manifest, respectively. Only one section is permitted.
+
+For libraries:
+
+```toml
+[lib]
+start = "0x..." # Optional
+```
+
+The `start` field specifies the MAST root of a function that must be called to initialize the library before
+any of its other exports are called. This function must be called for each new context in which the library
+is used, and must only be called once.
+
+For programs:
+
+```toml
+[bin]
+entrypoint = "0x..."
+```
+
+The `entrypoint` field specifies the MAST root of a function that will be called as the entrypoint of the
+program, e.g. `_start`. An entrypoint is required, and the root specified _must_ be contained within the
+primary MAST of the package (i.e. not in a dependency).
+
+#### Read-Only Data/Data Segments
+
+If the package has static data it expects to be present in linear memory when any of the code it contains starts
+executing, it should be specified using the `[rodata]` section:
+
+```toml
+[rodata]
+# Write the contents of `assets/foo.json` into linear memory at address 0x0,
+0x0 = { path = "assets/foo.json", encoding = "bytes" }
+0xDEADBEEF = { content = "0x...", encoding = "words" }
+```
+
+In this example, two segments will be written, with slightly different semantics:
+
+* The contents of `assets/foo.json` will be written into linear memory starting at address `0x0`, treating the
+raw binary data as if it is laid out in byte-addressable memory, and encoding it as data in word-addressable memory
+by taking each 4 byte chunk and encoding it as a single field element.
+
+* The data represented by the hex-encoded string in the `content` field will be written into linear memory
+starting at address `0xDEADBEEF`, treating the decoded data as if it is already in the canonical word-addressable memory
+representation. This means that every 8 bytes will be interpreted as the canonical little-endian byte representation
+of a single field element.
+
+In both cases, the data written into linear memory will be padded with zeros to the nearest word boundary. For example,
+if you have 13 elements worth of data (3 words + 1 element), then the actual segment size will be 16 elements, with
+the last 3 all zeroes.
+
+Segment overlap is not permitted, except when the overlap only occurs due to padding. Segments are written in increasing
+address order, so padding bytes will never overwrite non-padding bytes.
+
+> [!IMPORTANT]
+> The `[rodata]` section is only used for creating a package from a directory.
+> It will never be present in a manifest extracted from a package, because the
+> rodata will have been emitted in canonical on-disk form in the `rodata` directory
```

Comment:
It serves an important purpose (primarily crucial for those defining packages by hand). The file on disk must be in canonical form, encoded such that it can be directly copied into Miden's word-addressable memory. It isn't sufficient to, as an example, take some file containing a bunch of string constants, and put it in the `rodata` directory. That file must be encoded as some number of words, in the canonical binary representation for field elements. For data intended for a byte-addressable address space, this is done by slicing the data into 4 byte chunks, encoding each chunk as a field element, and then converting the field element into its canonical binary representation.

If you already have your data in canonical form, then yes, you don't need the `[rodata]` section (and in fact, you can omit it in favor of placing the files in the `rodata` directory instead). However, if you are creating a package, and what you have is that file of string constants I mentioned above, you're in trouble - there is no tooling that will do the conversion for you, and doing it yourself is error prone. The `[rodata]` section is specifically aimed at automating this conversion for you using the package tooling. The result is, you can maintain the sources for a package in the human-friendly form, but easily produce the binary package with our tools.

To be clear, the `[rodata]` section is only present when the manifest is being written by hand. When extracting a package to disk, regardless of how that package was defined, the manifest elides that section, and the canonical form is dumped to the `rodata` directory.

Reply:
> We could introduce a separate command in our cargo extension or even a new conversion tool for generating the binary files in canonical form.

I think the risk here is _knowing_ that you have to do this, let alone how, or that we provide a tool to do it. I suspect many users aren't even going to realize/understand that how they normally think about binary data simply does not translate to Miden's memory model. It's not even feasible for us to warn them about this, because we don't know anything about the data and how it is intended to be used.

> I think this adds an extra complexity for us and our users. I prefer the manifest to be the same before/after processing.

The complexity exists one way or the other. Either we push it on to our users, or we provide a solution for it. In our case, we already solve this problem for memory in general in the compiler, so we are best equipped to handle this for them, and to ensure its done correctly.

The tradeoff of the encoded manifest being different than the hand-written one is just that, a tradeoff. To be clear though, there is no technical reason that we can't extract to a form that is identical to that created by the user, just that it requires storing more metadata in the encoded package to do so, so that we can recover the original file names, and an additional post-processing step would be required to convert the data back to its original form.

The reason I chose not to do that here is twofold: First, it is not useful to the primary consumer of a package, the VM. Second, it is additional data to store for the case in which someone wishes to extract a package into its original source form. That doesn't mean it isn't useful, just that it felt like optimizing for an edge case.